libqpdf/QPDFWriter.cc

```diff
@@ -3154,3 +3199,88 @@ QPDFWriter::writeStandard()
         assert(this->md5_pipeline == 0);
     }
 }
+
+void
+QPDFWriter::writePCLm()
+{
+    if (this->deterministic_id)
+    {
+        pushMD5Pipeline();
+    }
+
+    // Start writing
+
+    writePCLmHeader();
+    writeString(this->extra_header_text);
+
+    // Image transform stream content for page strip images.
+    // Each of this new stream has to come after every page image
+    // strip written in the pclm file.
+    std::string image_transform_content = "q /image Do q\n";
+
+    // enqueue all pages first
+    std::vector<QPDFObjectHandle> all = this->pdf.getAllPages();
+    for (std::vector<QPDFObjectHandle>::iterator iter = all.begin();
+            iter != all.end(); ++iter)
+    {
+        // enqueue page
+        enqueueObject(*iter);
+
+        // enqueue page contents stream
+        enqueueObject((*iter).getKey("/Contents"));
+
+        // enqueue all the strips for each page
+        QPDFObjectHandle strips = (*iter).getKey("/Resources").getKey("/XObject");
+        std::set<std::string> keys = strips.getKeys();
+        for (std::set<std::string>::iterator image = keys.begin();
+                image != keys.end(); ++image)
+        {
+            enqueueObject(strips.getKey(*image));
+            enqueueObject(QPDFObjectHandle::newStream(&pdf, image_transform_content));
+        }
+    }
+
+    // Put root in queue.
+    QPDFObjectHandle trailer = getTrimmedTrailer();
+    enqueueObject(trailer.getKey("/Root"));
+
+    // Now start walking queue, output each object
+    while (this->object_queue.size())
+    {
+	QPDFObjectHandle cur_object = this->object_queue.front();
+	this->object_queue.pop_front();
+	writeObject(cur_object);
+    }
+
+    // Write out the encryption dictionary, if any
+    if (this->encrypted)
+    {
+	writeEncryptionDictionary();
```

Comment:
@jberkenbilt @tillkamppeter
It does not make sense to me to have encrypted PCLm files, since PCLm would not support many of the encryption keys. It may be visible in a standard PDF editor, but will not be able to print in a PCLm supported printer. Also, I just got to know that QDF mode is also not supported by the printer. Both of these phenomenon imply that neither a PCLm with QDF mode enabled nor PCLm with encryption will be a valid PCLm since both of them will be unable to print on a printer. So what do you suggest, do we drop these two methods for PCLm?